In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [2]:
df = pd.read_csv('DataScience_salaries_2025.csv')

df.shape
print(df['experience_level'].unique()) # cardinality of 4
print(df['company_size'].unique()) # cardinality of 3
len(df['job_title'].unique()) 

['MI' 'SE' 'EN' 'EX']
['M' 'L' 'S']


317

In [3]:
len(df[df['employee_residence'] == "US"])/len(df)

0.9005630522345802

In [4]:
print(df.duplicated().sum())                 # count of fully-identical rows
df[df.duplicated(keep=False)].sort_values(list(df.columns)).head(20)

46960


,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
93444,2021,EN,FT,Data Scientist,90000,USD,90000,US,100,US,S
93519,2021,EN,FT,Data Scientist,90000,USD,90000,US,100,US,S
93359,2021,MI,FT,Data Engineer,200000,USD,200000,US,100,US,L
93439,2021,MI,FT,Data Engineer,200000,USD,200000,US,100,US,L
93515,2021,MI,FT,Data Scientist,76760,EUR,90734,DE,50,DE,L
93553,2021,MI,FT,Data Scientist,76760,EUR,90734,DE,50,DE,L
90666,2021,MI,FT,Data Scientist,150000,USD,150000,US,100,US,M
93385,2021,MI,FT,Data Scientist,150000,USD,150000,US,100,US,M
91902,2022,EN,FT,Data Analyst,50000,USD,50000,US,50,US,L
91919,2022,EN,FT,Data Analyst,50000,USD,50000,US,50,US,L


In [5]:
df.head()

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
0,2025,MI,FT,Research Scientist,208000,USD,208000,US,0,US,M
1,2025,MI,FT,Research Scientist,147000,USD,147000,US,0,US,M
2,2025,SE,FT,Research Scientist,173000,USD,173000,US,0,US,M
3,2025,SE,FT,Research Scientist,117000,USD,117000,US,0,US,M
4,2025,MI,FT,AI Engineer,100000,USD,100000,US,100,US,M


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93597 entries, 0 to 93596
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           93597 non-null  int64 
 1   experience_level    93597 non-null  object
 2   employment_type     93597 non-null  object
 3   job_title           93597 non-null  object
 4   salary              93597 non-null  int64 
 5   salary_currency     93597 non-null  object
 6   salary_in_usd       93597 non-null  int64 
 7   employee_residence  93597 non-null  object
 8   remote_ratio        93597 non-null  int64 
 9   company_location    93597 non-null  object
 10  company_size        93597 non-null  object
dtypes: int64(4), object(7)
memory usage: 7.9+ MB


This below cell is for my own curiosity, but these salaries seem really, really high

In [7]:
#df=df[(df["experience_level"] == "EN") & (df["employment_type"] == "FT") & (df['employee_residence'] == 'US')]
#df.drop(columns=['experience_level','employment_type','employee_residence'])

there are no null values within this dataset, it's from kaggle so this isn't too surprising

In [8]:
df.isnull().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

In [16]:
entrydf = df[df["experience_level"] == "EN"]

entrydf['job_title'].value_counts().head(30)

job_title
Data Analyst                       3434
Analyst                             743
Data Scientist                      640
Data Engineer                       454
Software Engineer                   421
Machine Learning Engineer           354
Engineer                            342
Research Scientist                  281
Research Analyst                    211
Business Intelligence Analyst       197
Business Analyst                    101
Data Management Analyst              88
Research Engineer                    85
Data Specialist                      83
Applied Scientist                    66
Data Quality Analyst                 61
Product Analyst                      52
BI Analyst                           50
Associate                            49
AI Engineer                          49
Consultant                           48
Data Governance Analyst              46
Insight Analyst                      46
Developer                            44
Analytics Engineer            

Vast majority of the jobs listed are from the US(90%)

In [10]:
len(df[df['employee_residence'] == "US"])/len(df)

0.9005630522345802

Job title shows high cardinality. amd as expected favors data scientist/engineer/analyst roles. It's surprising to see Software Engineer in 3rd place, but it could be that Software Engineers do a lot of data analyst adjunct work

In [11]:
(len(df[df['job_title'] == "Data Analyst"]))/df['job_title'].value_counts().sum()


np.float64(0.09702234045963011)

There are 46960 duplicates in this dataset. While this number may seem large, identical rows are expected with this dataset: there are only 4 experience levels x ~15 most common titles x 3 company sizes x 90% of responders are from the US x salaries are rounded off. The number of distinct possible profiles within this dataset is somewhat low, and people who have the same job at the same company will naturally have the same(or similar) salaries. There's no unique ID for respondent, so i'll keep this for descriptive views and I'll report the test with and without duplicates

In [12]:
df.duplicated().sum()

np.int64(46960)

In [13]:
df.duplicated(keep=False).sum()  # 61,092  — rows that share their signature with ≥1 other (65%)
df.value_counts().max()          # 186 - the frequency of the most common row
df.value_counts().head(3)        # see which rows those are

work_year  experience_level  employment_type  job_title       salary  salary_currency  salary_in_usd  employee_residence  remote_ratio  company_location  company_size
2025       SE                FT               Data Scientist  160000  USD              160000         US                  100           US                M               186
2024       SE                FT               Data Scientist  160000  USD              160000         US                  100           US                M               182
2025       SE                FT               Data Scientist  110000  USD              110000         US                  100           US                M               180
Name: count, dtype: int64

This shows salaries for different roles across 

In [14]:
dd = df.drop_duplicates()
roles = ['Data Analyst','Data Engineer','Data Scientist','Machine Learning Engineer', "Software Engineer"]
dd[dd.job_title.isin(roles)].groupby(['experience_level','job_title'])['salary_in_usd'].median().unstack()

job_title,Data Analyst,Data Engineer,Data Scientist,Machine Learning Engineer,Software Engineer
experience_level,,,,,
EN,81000.0,84000.0,86520.0,129600.0,111550.0
EX,114880.0,200000.0,192018.5,215200.0,229993.0
MI,88000.0,116047.0,121800.0,154500.0,155000.0
SE,116700.0,151500.0,161000.0,189360.0,182312.0


In [15]:
import duckdb
con = duckdb.connect()
con.register("salaries", df)   # df = the full 93,597-row frame

# Check 1: row count matches the file
con.sql("SELECT COUNT(*) AS n FROM salaries").show()   # expect 93597

# Check 2: reproduce a known median.
# Your verified SE Data Analyst median ($116,700) is the DEDUP'd figure,
# so dedup here or it won't match.
con.sql("""
  SELECT MEDIAN(salary_in_usd) AS med, COUNT(*) AS n
  FROM (SELECT DISTINCT * FROM salaries)
  WHERE experience_level = 'SE' AND job_title = 'Data Analyst'
""").show()   # expect med ≈ 116700

ModuleNotFoundError: No module named 'duckdb'

In [ ]:
con.sql("""
  SELECT job_title, experience_level,COUNT(*) AS n, MEDIAN(salary_in_usd) AS med
  FROM (SELECT DISTINCT * FROM salaries)
  WHERE job_title = 'Data Analyst' OR job_title = 'Data Scientist' OR job_title = 'Data Engineer' OR job_title = 'Machine Learning Engineer' GROUP BY experience_level, job_title ORDER BY med DESC
""").show()   

┌───────────────────────────┬──────────────────┬───────┬──────────┐
│         job_title         │ experience_level │   n   │   med    │
│          varchar          │     varchar      │ int64 │  double  │
├───────────────────────────┼──────────────────┼───────┼──────────┤
│ Machine Learning Engineer │ EX               │    43 │ 215200.0 │
│ Data Engineer             │ EX               │   264 │ 200000.0 │
│ Data Scientist            │ EX               │   156 │ 192018.5 │
│ Machine Learning Engineer │ SE               │  1785 │ 189360.0 │
│ Data Scientist            │ SE               │  3000 │ 161000.0 │
│ Machine Learning Engineer │ MI               │   711 │ 154500.0 │
│ Data Engineer             │ SE               │  2796 │ 151500.0 │
│ Machine Learning Engineer │ EN               │   149 │ 129600.0 │
│ Data Scientist            │ MI               │  1835 │ 121800.0 │
│ Data Analyst              │ SE               │  1845 │ 116700.0 │
│ Data Engineer             │ MI               │

In [ ]:
con.sql("""
WITH medians AS (
    SELECT 
        experience_level,
        MEDIAN(CASE WHEN job_title = 'Data Analyst' THEN salary_in_usd END) AS analyst_med,
        MEDIAN(CASE WHEN job_title = 'Data Scientist' THEN salary_in_usd END) AS scientist_med,
        MEDIAN(CASE WHEN job_title = 'Data Engineer' THEN salary_in_usd END) AS engineer_med,
        MEDIAN(CASE WHEN job_title = 'Machine Learning Engineer' THEN salary_in_usd END) AS ml_med
    FROM (SELECT DISTINCT * FROM salaries)
    GROUP BY experience_level
)

SELECT 
    experience_level,
    analyst_med,
    analyst_med - scientist_med AS scientist_median_diff,
    analyst_med - engineer_med AS engineer_median_diff,
    analyst_med - ml_med AS ml_median_diff
FROM medians ORDER BY analyst_med;
""").show()

┌──────────────────┬─────────────┬───────────────────────┬──────────────────────┬────────────────┐
│ experience_level │ analyst_med │ scientist_median_diff │ engineer_median_diff │ ml_median_diff │
│     varchar      │   double    │        double         │        double        │     double     │
├──────────────────┼─────────────┼───────────────────────┼──────────────────────┼────────────────┤
│ EN               │     81000.0 │               -5520.0 │              -3000.0 │       -48600.0 │
│ MI               │     88000.0 │              -33800.0 │             -28047.0 │       -66500.0 │
│ EX               │    114880.0 │              -77138.5 │             -85120.0 │      -100320.0 │
│ SE               │    116700.0 │              -44300.0 │             -34800.0 │       -72660.0 │
└──────────────────┴─────────────┴───────────────────────┴──────────────────────┴────────────────┘



In [ ]:
con.sql("""
WITH dupeminus AS (
    SELECT DISTINCT *
    FROM salaries
)
SELECT
    median(CASE WHEN job_title = 'Data Analyst'
                 AND experience_level = 'SE'
                THEN salary_in_usd END) AS analyst_median,

    COUNT(CASE WHEN job_title = 'Data Analyst'
                AND experience_level = 'SE'
               THEN 1 END) AS analyst_count,

    median(CASE WHEN job_title = 'Data Scientist'
                 AND experience_level = 'SE'
                THEN salary_in_usd END) AS scientist_median,

    COUNT(CASE WHEN job_title = 'Data Scientist'
                AND experience_level = 'SE'
               THEN 1 END) AS scientist_count
FROM dupeminus
""").show()

┌────────────────┬───────────────┬──────────────────┬─────────────────┐
│ analyst_median │ analyst_count │ scientist_median │ scientist_count │
│     double     │     int64     │      double      │      int64      │
├────────────────┼───────────────┼──────────────────┼─────────────────┤
│       116700.0 │          1845 │         161000.0 │            3000 │
└────────────────┴───────────────┴──────────────────┴─────────────────┘

